### PR TITLE
Serialize side table and integrate `verify`

### PR DESCRIPTION
**Pull request by zhouwfang**
#46

### CODE REVIEW

File: crates/interpreter/src/side_table.rs [hunk 1]
```diff
@@ -69,7 +71,7 @@ impl<'m> Metadata<'m> {
         unsafe {
             core::slice::from_raw_parts(
                 self.0[10 ..].as_ptr() as *const BranchTableEntry,
-                self.0.len() / entry_size,
+                self.0.len() / entry_size - 1,
```
Comment (ia0) [inline crates/interpreter/src/side_table.rs:74]: Why is there a `- 1` here?

File: crates/interpreter/src/side_table.rs [hunk 2]
```diff
@@ -86,6 +88,32 @@ pub struct MetadataEntry {
...
+pub fn serialize(side_table: &[MetadataEntry]) -> Result<Vec<u8>, Error> {
+    let mut res = vec![];
+    res.extend_from_slice(&(side_table.len() as u16).to_le_bytes());
+    let mut index = 0;
+    res.extend_from_slice(&(index as u16).to_le_bytes());
+    for entry in side_table {
+        index =
+            try_from::<u16>("index of MetadataEntry", index + 10 + 6 * entry.branch_table.len())?
+                as usize;
+        res.extend_from_slice(&(index as u16).to_le_bytes());
```
Comment (ia0) [inline crates/interpreter/src/side_table.rs:100]: Looking at this, we could actually have a `push_as<T: TryFrom<usize>>(dst: &mut Vec<u8>, src: usize, msg: &str)` helper instead of `try_from`. This should make this function much easier to read.

**Review by ia0 — COMMENTED** [review]
Thanks! I think the issues come from the fact that the side-table format is not correct yet.

File: crates/interpreter/src/module.rs [hunk 3]
```diff
@@ -73,6 +67,15 @@ impl<'m> Module<'m> {
                 module.types.push(parser.parse_functype().into_ok());
             }
         }
+        if let Some(mut parser) = module.section(SectionId::Custom) {
+            let section_name = "wasefire-sidetable";
+            if parser.parse_name().into_ok() == section_name {
+                let side_table_view = SideTableView::new(parser.save()).unwrap();
+                module.side_table = side_table_view;
+                let len = 3 + section_name.as_bytes().len() + parser.save().len();
+                module.binary = &binary[8 + len ..];
```
[the remark this answers is not in the document]
Comment (zhouwfang) [reply crates/interpreter/src/module.rs:76]: Are you sure? I checked the debugging logs and it does not seem the case.

File: crates/interpreter/src/module.rs [hunk 4]
```diff
@@ -178,8 +178,11 @@ impl<'m> Module<'m> {
     }
 
     pub(crate) fn func(&self, x: FuncIdx) -> (Parser<'m>, &'m [BranchTableEntry]) {
-        let MetadataEntry { parser_range, branch_table, .. } = &self.side_table[x as usize];
-        (unsafe { Parser::new(&self.binary[parser_range.clone()]) }, branch_table)
+        let metadata = Box::leak(Box::new(self.side_table.metadata(x as usize)));
+        (
+            unsafe { Parser::new(&self.binary[metadata.parser_range().clone()]) },
```
[the remark this answers is not in the document]
Comment (zhouwfang) [reply crates/interpreter/src/module.rs:183]: It does return a `Range`. I removed the `clone()`.

File: crates/interpreter/src/module.rs [hunk 5]
```diff
@@ -178,8 +178,11 @@ impl<'m> Module<'m> {
     }
 
     pub(crate) fn func(&self, x: FuncIdx) -> (Parser<'m>, &'m [BranchTableEntry]) {
-        let MetadataEntry { parser_range, branch_table, .. } = &self.side_table[x as usize];
-        (unsafe { Parser::new(&self.binary[parser_range.clone()]) }, branch_table)
+        let metadata = Box::leak(Box::new(self.side_table.metadata(x as usize)));
```
[the remark this answers is not in the document]
Comment (zhouwfang) [reply crates/interpreter/src/module.rs:181]: `metadata()` returns a temporary value.

File: crates/interpreter/src/side_table.rs (hunk 2 above)
Comment (zhouwfang) [reply crates/interpreter/src/side_table.rs:100]: T should also satisfy `ToBytes` for the `to_le_bytes()` to work. But `ToBytes` is not implemented for `u16` and `u32`.

File: crates/interpreter/src/module.rs [hunk 6]
```diff
@@ -73,6 +67,15 @@ impl<'m> Module<'m> {
                 module.types.push(parser.parse_functype().into_ok());
             }
         }
+        if let Some(mut parser) = module.section(SectionId::Custom) {
```
[the remark this answers is not in the document]
Comment (zhouwfang) [reply crates/interpreter/src/module.rs:70]: Do you mean not calling `section()` but writing a new `first/sidetable_section()`?

File: crates/interpreter/src/format.rs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+pub fn leb128(wasm: &mut Vec<u8>, mut x: usize) {
+    assert!(x <= u32::MAX as usize);
+    while x > 127 {
+        wasm.push(0x80 | (x & 0x7f) as u8);
+        x >>= 7;
+    }
+    wasm.push(x as u8);
+}
+
+pub fn section(wasm: &mut Vec<u8>, id: u8, content: &[u8], name: &str) {
```
[the remark this answers is not in the document]
Comment (zhouwfang) [reply crates/interpreter/src/format.rs:26]: Can I rename it to be `custom_section`?

[the remark this answers is not in the document]
Comment (ia0) [reply crates/interpreter/src/format.rs:26]: Yes. In that case, let's move `name` before `content` since that's the order in the binary format.

File: crates/interpreter/src/format.rs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    assert!(id <= 12);
+    wasm.push(id);
+    let name_bytes = name.as_bytes();
+    leb128(wasm, 1 + name_bytes.len() + content.len());
```
Comment (ia0) [inline crates/interpreter/src/format.rs:30]: This assumes `name.len() < 128`. We should assert it.

File: crates/interpreter/src/format.rs [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        wasm.push(0x80 | (x & 0x7f) as u8);
+        x >>= 7;
+    }
+    wasm.push(x as u8);
+}
+
+pub fn section(wasm: &mut Vec<u8>, id: u8, content: &[u8], name: &str) {
+    assert!(id <= 12);
+    wasm.push(id);
+    let name_bytes = name.as_bytes();
```
Comment (ia0) [inline crates/interpreter/src/format.rs:29]: No need for this variable: `name.len() == name.as_bytes().len()`.

File: crates/interpreter/src/module.rs [hunk 10]
```diff
@@ -20,17 +20,14 @@ use crate::parser::{SkipData, SkipElem};
 use crate::side_table::*;
 use crate::syntax::*;
 use crate::toctou::*;
-use crate::valid::prepare;
 use crate::*;
 
 /// Valid module.
 #[derive(Debug, Default)]
```
Comment (ia0) [inline crates/interpreter/src/module.rs:26]: Let's merge `dev/fast-interp` to get rid of this `Default`.

File: crates/interpreter/src/module.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ia0) [reply crates/interpreter/src/module.rs:74]: We should use `Verify::parse_side_table(&mut Parser::new(module.binary))`. Or better, we should move the body of this function into `parser.rs` returning `MResult`, because in `valid.rs` we use a `Parser<Check>` but here in `module.rs` we use a `Parser<Use>`.

File: crates/interpreter/src/module.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ia0) [reply crates/interpreter/src/module.rs:80]: Custom sections are skipped by the `if actual_id == SectionId::Custom` test in `Module::section()`. Not sure what you mean by "not skipped".

File: crates/interpreter/src/module.rs [hunk 11]
```diff
@@ -109,9 +112,6 @@ impl<'m> Module<'m> {
             }
             let actual_id = parser.parse_section_id().into_ok();
             let section = parser.split_section().into_ok();
-            if actual_id == SectionId::Custom {
```
Comment (ia0) [inline crates/interpreter/src/module.rs:112]: Why are you removing this?

File: crates/interpreter/src/module.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ia0) [reply crates/interpreter/src/module.rs:185]: Ah the problem is the type of `Metadata::branch_table()`. It should return `&'m [BranchTableEntry]`.

File: crates/interpreter/src/side_table.rs [hunk 12]
```diff
@@ -69,7 +71,7 @@ impl<'m> Metadata<'m> {
         unsafe {
             core::slice::from_raw_parts(
                 self.0[10 ..].as_ptr() as *const BranchTableEntry,
-                self.0.len() / entry_size,
+                (self.0.len() - 10) / entry_size,
```
Comment (ia0) [inline crates/interpreter/src/side_table.rs:74]: Let's improve this by using an intermediate variable:

```rust
let branch_table = &self.0[10 ..];
assert_eq!(branch_table.len() % entry_size, 0);
from_raw_parts(branch_table.as_ptr().cast(), branch_table.len() / entry_size)
```

File: crates/interpreter/src/side_table.rs (hunk 2 above)
Comment (ia0) [reply crates/interpreter/src/side_table.rs:100]: I see. There's actually no trait for that sadly. `ToBytes` is for SIMD. We could create our own but then it's not worth it. Let's leave it like that.

Comment (ia0) [inline crates/interpreter/src/side_table.rs:100]: We should actually have `index: u16`. The casts are at the wrong places. It should be:

```rust
    let mut index = 0u16;
    res.extend_from_slice(&index.to_le_bytes());
    for entry in side_table {
        index = try_from::<u16>(
            "index of MetadataEntry",
            index as usize + 10 + 6 * entry.branch_table.len(),
        )?;
        res.extend_from_slice(&index.to_le_bytes());
    }
```

**Comment by zhouwfang** [comment]
Now that `valid.rs` becomes generic over `ValidMode`, I wonder how I could print debugging logs for generic objects. For example, I'm not sure how to print the next [branch table](https://github.com/zhouwfang/wasefire/blob/bc7d5762ddef7ed46287819ebf7e8e52d5330c2b/crates/interpreter/src/valid.rs#L314). I added `Debug` at some places, but there still seems something missing. Could you show me the code for it so that I could debug over the weekends? Thanks!

**Comment by ia0** [comment]
> Now that `valid.rs` becomes generic over `ValidMode`, I wonder how I could print debugging logs for generic objects.

I've pushed a commit for that (called `debug`). For generic contexts you need the `Debug` bound on the associated types (not just on the final types used for those associated types in the trait implementations).

File: crates/interpreter/src/valid.rs [hunk 13]
```diff
@@ -179,14 +185,18 @@ impl<'m> BranchTableApi<'m> for MetadataView<'m> {
     fn stitch_branch(
         &mut self, source: SideTableBranch<'m>, target: SideTableBranch<'m>,
     ) -> CheckResult {
-        check(source == target)
+        check(
+            source.parser == target.parser
+                && source.result == target.result
+                && source.branch_table == target.branch_table
+                && source.stack <= target.stack,
```
Comment (zhouwfang) [inline crates/interpreter/src/valid.rs:192]: @ia0 This is consistent with [the logic](https://github.com/zhouwfang/wasefire/blob/b1e4cb9884707897da37ee5e2ca1db322cb4b2c2/crates/interpreter/src/valid.rs#L1146-L1148) in `pop_cnt()`. To be completely rigorous, we would need to add `pop_cnt == 0`, but it does not seem straightforward to obtain `pop_cnt` whenever `stitch_branch()` is called.

Comment (zhouwfang) [inline crates/interpreter/src/valid.rs:192]: @ia0 This is a necessary condition of [the logic](https://github.com/zhouwfang/wasefire/blob/b1e4cb9884707897da37ee5e2ca1db322cb4b2c2/crates/interpreter/src/valid.rs#L1146-L1148) in `pop_cnt()`. The actual equivalent condition should be `source.stack == target.stack || (pop_cnt == 0 && source.stack < target.stack)`, but it does not seem straightforward to obtain `pop_cnt` whenever `stitch_branch()` is called.

Comment (zhouwfang) [reply crates/interpreter/src/valid.rs:192]: Update: we can store `pop_cnt` as `Option<usize>` in `SideTableBranch`.

**Comment by zhouwfang** [comment]
The only test that fails in the test suite now is [the huge br_table](https://github.com/WebAssembly/spec/blob/c280b356e0da830ce0b7c636c34ea684f5d1b728/test/core/br_table.wast#L107) because of ["index of MetadataEntry overflow"](https://github.com/zhouwfang/wasefire/blob/3b732824890cbdac9089da32f659f3d80279563c/crates/interpreter/src/side_table.rs#L86-L89) (the index is 97252). Do we want to change it to u32?

I suspect the error in `./scripts/hwci.sh` is due to the same reason.

File: crates/interpreter/src/valid.rs (hunk 13 above)
Comment (ia0) [reply crates/interpreter/src/valid.rs:199]: I'll need to look into this when I get time. I think the whole situation around pop_cnt feels wrong, but maybe I'm missing something. I've added some TODOs to not lose track of this. This should not block this PR.

**Review by ia0 — APPROVED** [review]
Thanks! That's a huge step, but I'm not convinced about the pop_cnt story yet. I'll need to look into it. Ideally we should do some differential fuzzing, I don't trust the test suite to provide good coverage.

**Comment by ia0** [comment]
> The only test that fails in the test suite now is [the huge br_table](https://github.com/WebAssembly/spec/blob/c280b356e0da830ce0b7c636c34ea684f5d1b728/test/core/br_table.wast#L107) because of ["index of MetadataEntry overflow"](https://github.com/zhouwfang/wasefire/blob/3b732824890cbdac9089da32f659f3d80279563c/crates/interpreter/src/side_table.rs#L86-L89) (the index is 97252). Do we want to change it to u32?
> 
> I suspect the error in `./scripts/hwci.sh` is due to the same reason (but I'm not sure how to check that).

I'm not sure we want to support such corner-cases. For now I'm ignoring all 149 tests in `br_table`, but maybe some of them we shouldn't. We can deal with that later.

If hwci is still failing, I'll take a look and work around it to merge. It might be because of the pop_cnt story which might be wrong and is not caught by the test suite.

**Review by ia0 — APPROVED** [review]

File: crates/interpreter/src/valid.rs [hunk 14]
```diff
@@ -1122,6 +1149,8 @@ fn delta(
 
 fn pop_cnt(source: SideTableBranch, target: SideTableBranch) -> MResult<u32, Check> {
     let source = source.stack;
+    // TODO(dev/fast-interp): Figure out why we can't simply source.stack - target.stack and
+    // document it.
```
Comment (zhouwfang) [inline crates/interpreter/src/valid.rs:1153]: @ia0 I thought we shouldn't always expect `source.stack` to be `>= target.stack`. One example is the [as-if-cond](https://github.com/WebAssembly/spec/blob/2e2e3a9e471d162ce317b6d1153e8ff8b9945218/test/core/br_if.wast#L108) in `br_if.wast`. At `Else`, we have ` source=SideTableBranch { parser: [65, 2, 5, 65, 3, 11, 11, 11], branch_table: 1, stack: 0, result: 0 }, target=SideTableBranch { parser: [65, 3, 11, 11, 11], branch_table: 3, stack: 1, result: 1 }`.

**Comment by zhouwfang** [comment]
> > The only test that fails in the test suite now is [the huge br_table](https://github.com/WebAssembly/spec/blob/c280b356e0da830ce0b7c636c34ea684f5d1b728/test/core/br_table.wast#L107) because of ["index of MetadataEntry overflow"](https://github.com/zhouwfang/wasefire/blob/3b732824890cbdac9089da32f659f3d80279563c/crates/interpreter/src/side_table.rs#L86-L89) (the index is 97252). Do we want to change it to u32?
> > I suspect the error in `./scripts/hwci.sh` is due to the same reason (but I'm not sure how to check that).
> 
> I'm not sure we want to support such corner-cases. For now I'm ignoring all 149 tests in `br_table`, but maybe some of them we shouldn't. We can deal with that later.
> 
> If hwci is still failing, I'll take a look and work around it to merge. It might be because of the pop_cnt story which might be wrong and is not caught by the test suite.

I had checked `br_table.wast` throughly, and [large](https://github.com/WebAssembly/spec/blob/c280b356e0da830ce0b7c636c34ea684f5d1b728/test/core/br_table.wast#L107) is only function that fails due to the `u16` overflow (after I commented out this function, the test suite passes).

`pop_cnt` should be correct for `hwci` because I completed its creation in `valid.rs` and usage in `exec.rs` long time ago, and there has been no issue with `hwci`. The `hwci` error should be due to `u16` overflow.
